quic/s2n-quic-core/src/sync/spsc/state.rs

```diff
@@ -0,0 +1,479 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{
+    primitive::{AtomicBool, AtomicUsize, AtomicWaker, Ordering},
+    Cell, ClosedError, Result, Slice,
+};
+use alloc::alloc::Layout;
+use cache_padded::CachePadded;
+use core::{
+    fmt,
+    marker::PhantomData,
+    ops::Deref,
+    panic::{RefUnwindSafe, UnwindSafe},
+    ptr::NonNull,
+};
+
+type Pair<'a, T> = super::Pair<Slice<'a, Cell<T>>>;
+
+const MINIMUM_CAPACITY: usize = 2;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Side {
+    Sender,
+    Receiver,
+}
+
+#[derive(Clone, Copy)]
+pub struct Cursor {
+    head: usize,
+    tail: usize,
+    capacity: usize,
+}
+
+impl fmt::Debug for Cursor {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Cursor")
+            .field("head", &self.head)
+            .field("tail", &self.tail)
+            .field("len", &self.len())
+            .field("capacity", &self.capacity())
+            .field("is_empty", &self.is_empty())
+            .field("is_full", &self.is_full())
+            .field("is_contiguous", &self.is_contiguous())
+            .finish()
+    }
+}
+
+impl Cursor {
+    #[inline]
+    fn new(capacity: usize) -> Self {
+        Self {
+            head: 0,
+            tail: 0,
+            capacity,
+        }
+    }
+
+    #[inline]
+    fn invariants(&self) {
+        unsafe {
+            assume!(self.capacity >= MINIMUM_CAPACITY);
+            assume!(self.head < self.capacity);
+            assume!(self.tail < self.capacity);
+            let len = count(self.head, self.tail, self.capacity);
+            assume!(len < self.capacity);
+        }
+    }
+
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.invariants();
+        self.capacity - 1
```

Comment:
Why `- 1`?

Reply:
The cursor management is based on the one in VecDeque, which requires that we add one and round up to the next power of two. It basically boils down to the fact that if you allow the caller to write the full capacity, then you can't distinguish between an empty and a full queue because the cursors are equal. So if you limit capacity to `total - 1`, you can assume that when cursors are equal, the queue is empty.